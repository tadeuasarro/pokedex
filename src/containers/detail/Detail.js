import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { resetDetail } from '../../actions/index';
import showPokemon from '../../api/show-pokemon';
import './detail.css';

const Detail = detail => {
  const dispatch = useDispatch();

  const {
    weight, height, id, types, sprite, name,
  } = detail.detail;

  let idStr = String(id);

  if (idStr.length === 1) {
    idStr = `00${id}`;
  } else if (idStr.length === 2) {
    idStr = `0${id}`;
  } else {
    idStr = id;
  }

  const handleReset = () => {
    dispatch(resetDetail());
  };

  const handleNext = () => {
    if (id < 151) {
      dispatch(showPokemon(id + 1));
    }
  };

  const handlePrevious = () => {
    if (id > 1) {
      dispatch(showPokemon(id - 1));
    }
  };

  return (
    <div className="pokemon-detail-outer-container">
      <div className="pokemon-detail-container">
        <button type="button" onClick={handleReset} className="reset-button">Home</button>
        <button type="button" onClick={handleNext} className="next-button">Next</button>
        <button type="button" onClick={handlePrevious} className="previous-button">Previous</button>
        <div className="pokemon-detail-info">
          <p className="detail-info">
            #
            {idStr}
            &ensp;
            {name.toUpperCase()}
          </p>
          <p className="detail-info">
            {
              types.map(type => (
                <span key={type.type.name}>
                  {type.type.name.toUpperCase()}
                  &ensp;
                </span>
              ))
            }
          </p>
          <p className="detail-info">
            Height
            &ensp;
            {height / 10}
            &ensp;
            m
          </p>
          <p>
            Weight
            &ensp;
            {weight / 10}
            &ensp;
            kg
          </p>
        </div>
        <div>
          <div className="pokemon-sprite" style={{ backgroundImage: `url(${sprite})` }}> </div>
        </div>
      </div>
    </div>
  );
};

Detail.propTypes = {
  detail: PropTypes.shape({
    weight: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
    types: PropTypes.instanceOf(Array),
    sprite: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }),
};

Detail.defaultProps = {
  detail: {},
};

export default Detail;
