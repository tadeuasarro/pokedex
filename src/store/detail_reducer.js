import { FETCH_DETAIL_ERROR, FETCH_DETAIL_PENDING, FETCH_DETAIL_SUCCESS } from '../actions/index';

const detailInitial = {
  pending: false,
  results: false,
  error: null,
};

export const detailReducer = (state = detailInitial, action) => {
  switch (action.type) {
    case FETCH_DETAIL_PENDING:
      return {
        ...state,
        pending: true,
      };
    case FETCH_DETAIL_SUCCESS:
      return {
        ...state,
        pending: false,
        results: action.detail,
      };
    case FETCH_DETAIL_ERROR:
      return {
        ...state,
        pending: false,
        error: action.error,
      };
    default:
      return state;
  }
};

export const getDetail = state => state.detail;
export const getDetailPending = state => state.pending;
export const getDetailError = state => state.error;
