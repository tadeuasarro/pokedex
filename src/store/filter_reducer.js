import { FETCH_TYPES_ERROR, FETCH_TYPES_PENDING, FETCH_TYPES_SUCCESS } from '../actions/index';

const detailInitial = {
  pending: false,
  results: false,
  error: null,
};

export const filterReducer = (state = detailInitial, action) => {
  switch (action.type) {
    case FETCH_TYPES_PENDING:
      return {
        ...state,
        pending: true,
      };
    case FETCH_TYPES_SUCCESS:
      return {
        ...state,
        pending: false,
        results: action.types,
      };
    case FETCH_TYPES_ERROR:
      return {
        ...state,
        pending: false,
        error: action.error,
      };
    default:
      return state;
  }
};

export const getTypes = state => state.types;
export const getTypesPending = state => state.pending;
export const getTypesError = state => state.error;
