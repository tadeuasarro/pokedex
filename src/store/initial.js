const pokemonInitial = {
  pending: false,
  results: [],
  error: null,
  detail: false,
};

export default pokemonInitial;
