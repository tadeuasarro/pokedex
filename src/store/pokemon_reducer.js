import { FETCH_POKEMON_PENDING, FETCH_POKEMON_SUCCESS, FETCH_POKEMON_ERROR } from '../actions/index';

const pokemonInitial = {
  pending: false,
  results: [],
  error: null,
};

export const pokemonReducer = (state = pokemonInitial, action) => {
  switch (action.type) {
    case FETCH_POKEMON_PENDING:
      return {
        ...state,
        pending: true,
      };
    case FETCH_POKEMON_SUCCESS:
      return {
        ...state,
        pokemon: action.pokemon,
      };
    case FETCH_POKEMON_ERROR:
      return {
        ...state,
        pending: false,
        error: action.error,
      };
    default:
      return state;
  }
};

export const getPokemon = state => state.results;
export const getPokemonPending = state => state.pending;
export const getPokemonError = state => state.error;
