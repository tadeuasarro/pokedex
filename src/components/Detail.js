import './detail.css';

const Detail = () => (
  <div className="pokemon-detail-outer-container">
    <div className="pokemon-detail-container">
      <div className="pokemon-detail-info">
        <p className="detail-info">#091 CLOYSTER</p>
        <p className="detail-info">BIVALVE POKÃ‰MON</p>
        <p className="detail-info">Height 4&apos;11&quot;</p>
        <p>Weight 292.1 lbs</p>
      </div>
      <div>Some pokemon img</div>
    </div>
  </div>
);

export default Detail;
