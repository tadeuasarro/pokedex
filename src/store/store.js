import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { pokemonReducer } from './pokemon_reducer';
import { detailReducer } from './detail_reducer';

const store = createStore(
  combineReducers({
    pokemon: pokemonReducer,
    detail: detailReducer,
  }),
  applyMiddleware(thunk),
);

export default store;
