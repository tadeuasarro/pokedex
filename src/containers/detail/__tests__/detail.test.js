import { render, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../../store/store';
import Detail from '../Detail';

const detail = {
  id: 1,
  height: 10,
  weight: 100,
  name: 'string',
  sprite: 'https://somestring.com',
  types: [
    {
      type: {
        name: 'grass',
      },
    },
  ],
};

describe('Rendering component', () => {
  it('creates an App component', () => {
    act(() => {
      render(
        <Provider store={store}>
          <Detail detail={detail} />
        </Provider>,
      );
    });
    // eslint-disable-next-line
    const element = screen.getByText('Home');
  });
});

describe('Display', () => {
  it('renders correctly', () => {
    const comp = renderer.create(
      <Provider store={store}>
        <Detail detail={detail} />
      </Provider>,
    ).toJSON();
    expect(comp).toMatchSnapshot();
  });
});
