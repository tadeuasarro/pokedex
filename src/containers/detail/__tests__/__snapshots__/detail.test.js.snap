// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correctly 1`] = `
<div
  className="pokemon-detail-outer-container"
>
  <div
    className="pokemon-detail-container"
  >
    <button
      className="reset-button"
      onClick={[Function]}
      type="button"
    >
      Home
    </button>
    <button
      className="next-button"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
    <button
      className="previous-button"
      onClick={[Function]}
      type="button"
    >
      Previous
    </button>
    <div
      className="pokemon-detail-info"
    >
      <p
        className="detail-info"
      >
        #
        001
         
        STRING
      </p>
      <p
        className="detail-info"
      >
        <span>
          GRASS
           
        </span>
      </p>
      <p
        className="detail-info"
      >
        Height  
        1
          m
      </p>
      <p>
        Weight  
        10
          kg
      </p>
    </div>
    <div>
      <div
        className="pokemon-sprite"
        style={
          Object {
            "backgroundImage": "url(https://somestring.com)",
          }
        }
      >
         
      </div>
    </div>
  </div>
</div>
`;
